openstack/datadog_checks/openstack/openstack.py

```diff
@@ -945,32 +926,58 @@ def get_all_servers(self, i_key, collect_all_tenants, filter_by_host=None):
                 except KeyError as e:
                     self.log.debug("Server: %s has already been removed from the cache", new_server['server_id'])
 
-        return self.server_details_by_id
+    def filter_excluded_servers(self):
+        proj_list = set([])
+        if self.exclude_server_id_rules:
+            # Filter out excluded servers
+            for exclude_id_rule in self.exclude_server_id_rules:
+                for server_id in self.server_details_by_id.keys():
+                    if re.match(exclude_id_rule, server_id):
+                        del self.server_details_by_id[server_id]
+
+        for _, server in self.server_details_by_id.iteritems():
+            proj_list.add(server.get('project_name'))
+
+        projects_filtered = pattern_filter(
+            proj_list,
+            whitelist=self.include_project_name_rules,
+            blacklist=self.exclude_project_name_rules
+        )
+
+        self.server_details_by_id = {
+                sid: server for (sid, server)
+                in self.server_details_by_id.iteritems()
+                if server.get('project_name') in projects_filtered
+        }
 
     def get_project_name_from_id(self, tenant_id):
-        url = "{0}/{1}/{2}/{3}".format(self.keystone_server_url, DEFAULT_KEYSTONE_API_VERSION, "projects", tenant_id)
+        url = "{}/{}/{}/{}".format(self.keystone_server_url, DEFAULT_KEYSTONE_API_VERSION, "projects", tenant_id)
         self.log.debug("Project URL is %s", url)
         headers = {'X-Auth-Token': self.get_auth_token()}
         try:
             r = self._make_request_with_auth_fallback(url, headers)
             return r['project']['name']
 
         except Exception as e:
-            self.warning('Unable to get project name: {0}'.format(str(e)))
-            raise e
+            self.warning('Unable to get project name: {}'.format(str(e)))
+            return "unknown_project"
```

Comment:
I don't think we should return an invalid project name. Instead, what about returning `None` and excluding this project from the list (with a warning) in the `get_all_servers` method?